libyul/AST.h

```diff
@@ -35,56 +35,81 @@ namespace solidity::yul
 
 using Type = YulString;
 
-struct TypedName { langutil::SourceLocation location; YulString name; Type type; };
+struct DebugData
+{
+	explicit DebugData(langutil::SourceLocation _location): location(std::move(_location)) {}
+	langutil::SourceLocation location;
+};
+
+inline std::shared_ptr<DebugData const> createDebugData(langutil::SourceLocation _location = {})
+{
+	return std::make_shared<DebugData const>(_location);
+}
+
+struct TypedName { std::shared_ptr<DebugData const> debugData; YulString name; Type type; };
 using TypedNameList = std::vector<TypedName>;
 
 /// Literal number or string (up to 32 bytes)
 enum class LiteralKind { Number, Boolean, String };
-struct Literal { langutil::SourceLocation location; LiteralKind kind; YulString value; Type type; };
+struct Literal { std::shared_ptr<DebugData const> debugData; LiteralKind kind; YulString value; Type type; };
 /// External / internal identifier or label reference
-struct Identifier { langutil::SourceLocation location; YulString name; };
+struct Identifier { std::shared_ptr<DebugData const> debugData; YulString name; };
 /// Assignment ("x := mload(20:u256)", expects push-1-expression on the right hand
 /// side and requires x to occupy exactly one stack slot.
 ///
 /// Multiple assignment ("x, y := f()"), where the left hand side variables each occupy
 /// a single stack slot and expects a single expression on the right hand returning
 /// the same amount of items as the number of variables.
-struct Assignment { langutil::SourceLocation location; std::vector<Identifier> variableNames; std::unique_ptr<Expression> value; };
-struct FunctionCall { langutil::SourceLocation location; Identifier functionName; std::vector<Expression> arguments; };
+struct Assignment { std::shared_ptr<DebugData const> debugData; std::vector<Identifier> variableNames; std::unique_ptr<Expression> value; };
```

Comment:
Regarding https://github.com/ethereum/solidity/pull/11294#discussion_r618411732: if sharing tiny pieces data between AST copies is actually beneficial then maybe we should also replace `unique_ptr`s and `vectors` with `shared_ptr`s too?
```suggestion
struct Assignment { std::shared_ptr<DebugData const> debugData; std::shared_ptr<std::vector<Identifier> const> variableNames; std::shared_ptr<Expression const> value; };
```

Not necessarily in this PR though. Just a general suggestion/question.

Reply:
I would do this in a separate PR, as I think it is not the topic for this one

Reply:
This could make code ugly, because you need to create a copy of the AST when you make a change inside it, and then nothing can be `const` while you are modifying it.